flamedisx/nest/lxe_sources.py

```diff
@@ -552,23 +555,262 @@ class nestWIMPSource(nestNRSource):
+
+        scale = time_bin_width / nu.year  # Convert from [per year] to [per time_bin_width]
+        scale *= energy_bin_width / nu.keV  # Convert from [per keV] to [per energy_bin_width]
+
+        rates_list = []
+        if modulation:
+            with ProcessPoolExecutor() as executor:
+                for time in times:
+                    rates_list.append(
+                        executor.submit(
+                            wr.rate_wimp_std,
+                            energy_values,
+                            mw=wimp_mass,
+                            sigma_nucleon=sigma,
+                            t=time,
+                        )
+                    )
+
+                rates_list = [future.result() for future in rates_list]
+        else:
+            rates = wr.rate_wimp_std(
+                energy_values,
+                mw=wimp_mass,
+                sigma_nucleon=sigma,
+            )
+            for _ in times:
+                rates_list.append(rates)
+
+        RATES = np.array(rates_list) * scale
+
+        hist = mh.Histdd.from_histogram(
+            RATES, [time_bin_edges, energy_bin_edges], axis_names=("time", "energy")
+        )
+
+        return hist
+
+
+@export
+class nestMigdalSource(nestERSource):
+
+    model_blocks = (fd_nest.WIMPEnergySpectrum,) + nestERSource.model_blocks[1:]
+
+    def __init__(
+        self,
+        *args,
+        wimp_mass=40,
+        sigma=1e-45,
+        fid_mass=1.0,
+        min_E=1e-2,
+        max_E=80.0,
+        n_energy_bins=800,
+        min_time="2019-09-01T08:28:00",
+        max_time="2020-09-01T08:28:00",
+        livetime=1.0,
+        n_time_bins=25,
+        modulation=True,
+        migdal_model="Cox",
+        **kwargs
+    ):
+        if migdal_model not in ["Ibe", "Cox", "Cox_dipole"]:
+            raise ValueError("Invalid Migdal model. Choose from 'Ibe', 'Cox', 'Cox_dipole'")
+  
+        if (livetime < 0 or livetime > 1) and not isinstance(livetime, float):
+            raise ValueError("Livetime should be a percentage between 0 and 1")
+      
+        if "detector" not in kwargs:
+            kwargs["detector"] = "default"
+
+        self.energy_hist = self.get_energy_hist(
+            wimp_mass=wimp_mass,
+            sigma=sigma,
+            min_E=min_E,
+            max_E=max_E,
+            n_energy_bins=n_energy_bins,
+            min_time=min_time,
+            max_time=max_time,
+            n_time_bins=n_time_bins,
+            modulation=modulation,
+            migdal_model=migdal_model,
+        )
+
+        scale = fid_mass * livetime
+        self.energy_hist *= scale
+
+        self.n_time_bins = len(self.energy_hist.bin_centers()[0])
+        e_centers = self.energy_hist.bin_centers()[1]
+        self.energies = fd.np_to_tf(e_centers)
+
+        self.array_columns = (("energy_spectrum", len(e_centers)),)
+
+        super().__init__(*args, **kwargs)
+
+    @staticmethod
+    def get_energy_hist(
```

Comment:
@lorenzomag can we avoid duplicating all this code?

Reply:
Hello, yes that is definitely needed. I just wasn't sure where to put the function. Shall I put it in some utils.py file? Or just keep it in there but out of the classes?

Reply:
Ok, I refactored the code and made a new module. The name of the module is just the name of the function for now, but we can change it to utils.py if needed, I guess